onnxruntime/core/providers/mkldnn/subgraph/mkldnn_custom_op.h

```diff
@@ -0,0 +1,283 @@
+      } else if (mklnode.name == "LRN") {
+        std::ostringstream os;
+        os << "LRN-" << mklnode.node_index << "-";
+        std::shared_ptr<MklLrn<T>> kernel;
+        kernel.reset(new MklLrn<T>(mklnode, params.provider, params.mkl_context));
+        kernel->ReadAttributes(params.attributes, os.str());
+        for (auto& index : mklnode.parent_nodes) {
+          kernel->parents_.push_back(context_.kernels[index]);
+        }
+        context_.kernels.push_back(kernel);
+      }
+    }
+  }
+
+  Status Compute(const ONNXRunTimeTensor* input_tensors, ONNXRunTimeTensor* const output_tensors) {
+    Status status;
+    for (auto& kernel : context_.kernels) {
+      status = kernel->Compute(input_tensors, output_tensors);
+      if (!status.IsOK())
+        break;
+    }
+    if (status.IsOK())
+      context_.stream->submit(context_.net);
+    return status;
+  }
+
+  ~SubgraphPrimitive() = default;
+
+ private:
+  struct SubgraphContext {
+    std::unique_ptr<mkldnn::stream> stream;
+    std::vector<mkldnn::primitive> net;
+    std::vector<std::shared_ptr<MklKernel>> kernels;
+
+    SubgraphContext() : stream(nullptr) {}
+  };
+
+  void Initialize(const ONNXRunTimeTensor* input_tensors) {
+    // Propagate mkldnn block format
+    // dst format of current node to src format of next node
+    mkldnn::memory::format source_format = mkldnn::memory::format::any;  // ONNXRuntime format
+    for (auto& kernel : context_.kernels) {
+      Status status = kernel->CreatePrimitives(input_tensors, cpu_engine_, context_.net, source_format);
+      if (status.IsOK())
+		kernel->ReorderWeights(input_tensors, cpu_engine_);
+    }
+  }
+
+  SubgraphContext context_;
+  mkldnn::engine& cpu_engine_;
+};
+
+// Pool which allows for reuse of MKLDNN Conv primitives which are expensive to instantiate.
+// To address thread safety, the primitives are stored in a map on thread local storage.
+template <typename T>
+class SubgraphPrimitivePool : public PrimitivePool<T> {
+ public:
+  static SubgraphPrimitive<T>* Get(const ONNXRunTimeTensor* input_tensors,
+                                   const std::string subgraph_key,
+                                   const SubgraphParams& params) {
+    SubgraphPrimitive<T>* primitive = dynamic_cast<SubgraphPrimitive<T>*>(
+        SubgraphPrimitivePool<T>::GetInstance().GetPrimitive(subgraph_key));
+
+    if (primitive == nullptr) {
+      auto subgraph_primitive = std::make_unique<SubgraphPrimitive<T>>(input_tensors, params);
+      primitive = subgraph_primitive.get();
+      SubgraphPrimitivePool<T>::GetInstance().SetPrimitive(subgraph_key, std::move(subgraph_primitive));
+    }
+    return primitive;
+  }
+
+ private:
+  SubgraphPrimitivePool() = default;
+  ~SubgraphPrimitivePool() = default;
+
+  static SubgraphPrimitivePool& GetInstance() {
+    static SubgraphPrimitivePool pool;
+    return pool;
+  }
+};
+}  // namespace
+
+template <typename T>
+class MkldnnCustomOp {
```

Comment:
i think we can find another name that is more intuitive and describes the functionality.
will discuss with @pranavsharma